src/nanoarrow/nanoarrow_testing.hpp

```diff
@@ -613,6 +618,458 @@ class TestingJSONWriter {
   };
 };
 
+#if defined(NANOARROW_TESTING_WITH_NLOHMANN_JSON)
+
+/// \brief Reader for the Arrow integration testing JSON format
+class TestingJSONReader {
```

Comment:
I would recommend to only put declarations in this header file and move implementations into a .cpp.

Reply:
Also, since this is entirely guarded by a `#ifdef`, perhaps it should go into separate files?